### PR TITLE
Implement SSE2 SetVector128 HW intrinsics - managed

### PR DESCRIPTION
**Pull request by 4creators**
FYI @CarolEidt @fiigii @tannergooding 

~~This is managed implementation only to get early feedback on implementation. Tests will be added a bit later today or actually tomorrow in my time zone.~~

### CODE REVIEW

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs [hunk 1]
```diff
@@ -971,47 +972,136 @@ public static class Sse2
         /// ___m128i _mm_set_epi8 (char e15, char e14, char e13, char e12, char e11, char e10, char e9, char e8, char e7, char e6, char e5, char e4, char e3, char e2, char e1, char e0)
         ///   HELPER
         /// </summary>
-        public static Vector128<sbyte> SetVector128(sbyte e15, sbyte e14, sbyte e13, sbyte e12, sbyte e11, sbyte e10, sbyte e9, sbyte e8, sbyte e7, sbyte e6, sbyte e5, sbyte e4, sbyte e3, sbyte e2, sbyte e1, sbyte e0)  => SetVector128(e15, e14, e13, e12, e11, e10, e9, e8, e7, e6, e5, e4, e3, e2, e1, e0);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<sbyte> SetVector128(sbyte e15, sbyte e14, sbyte e13, sbyte e12, sbyte e11, sbyte e10, sbyte e9, sbyte e8, sbyte e7, sbyte e6, sbyte e5, sbyte e4, sbyte e3, sbyte e2, sbyte e1, sbyte e0)
+        {
+            unsafe
```
Comment (tannergooding) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs]: nit: Put unsafe in the method name

Comment (tannergooding) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs]: (meaning `public static Vector128<sbyte> SetVector128`)

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs [hunk 2]
```diff
@@ -971,47 +972,136 @@ public static class Sse2
...
+            {
+                sbyte* buffer = stackalloc sbyte[16];
```
Comment (tannergooding) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs]: Please log a bug. We should profile this before release.

We should also test out if an unpack sequence (what clang is doing) or if a stack buffer of `int` is better (what gcc is doing)

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs [hunk 3]
```diff
@@ -971,47 +972,136 @@ public static class Sse2
...
+        }
         /// <summary>
         /// __m128i _mm_set_epi16 (short e7, short e6, short e5, short e4, short e3, short e2, short e1, short e0)
         ///   HELPER
         /// </summary>
-        public static Vector128<short> SetVector128(short e7, short e6, short e5, short e4, short e3, short e2, short e1, short e0) => SetVector128(e7, e6, e5, e4, e3, e2, e1, e0);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<short> SetVector128(short e7, short e6, short e5, short e4, short e3, short e2, short e1, short e0)
+        {
+            Vector128<int> vector1 = ConvertScalarToVector128Int32(e0);
```
Comment (tannergooding) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs]: Same, we should profile this and determine what the actual "best" sequence is.

**Comment by tannergooding** [comment]
@4creators, are you going to fixup `Sse.SetVector128` as well?

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs (hunk 3 above)
Comment (4creators) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs]: Thnx for review - I have created #16838 to track both performance validation and creation of performance tests for HW intrinsics.

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs (hunk 1 above)
Comment (4creators) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs]: Will fix.

**Comment by 4creators** [comment]
> are you going to fixup Sse.SetVector128 as well

@tannergooding I have fixed it already https://github.com/dotnet/coreclr/pull/16797/commits/d23d663dc06211ba08e2c3a4df694f1191754e5a and will address all review feedback soon.

**Comment by 4creators** [comment]
> are you going to fixup Sse.SetVector128 as well

Sorry, missed that it is SSE not SSE2 you are asking for. Sure, will do it while addressing feedback.

**Comment by 4creators** [comment]
Addressed code review feedback. Due to changes in #16797 this PR should be merged first, than I would need to change #16797 before merging and finally we could merge #16832

@CarolEidt @fiigii @tannergooding PTAL

**Comment by CarolEidt** [comment]
Would still like to get a review on this from @tannergooding before merging.

**Comment by 4creators** [comment]
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs [hunk 4]
```diff
@@ -467,7 +467,22 @@ public static class Sse
         /// <summary>
         /// __m128 _mm_set_ps (float e3, float e2, float e1, float e0)
         /// </summary>
-        public static Vector128<float> SetVector128(float e3, float e2, float e1, float e0) => SetVector128(e3, e2, e1, e0);
+        public static unsafe Vector128<float> SetVector128(float e3, float e2, float e1, float e0)
+        {
+            // TODO-CQ Optimize algorithm choice based on benchmarks
+
+            // Zero vector and load e2 et index 0
+            Vector128<float> e2Vector = LoadScalarVector128(&e2);
```
Comment (tannergooding) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs]: @CarolEidt, looking at more existing algorithms, it does look like we will want something that can allow users to use `float`/`double` in these methods (without cost).

It would seem like having an `UnsafeCast` method (or better name) would work well (assuming it also explicitly documents the value of the upper bits as "indeterminate"). -- Maybe we could update `Unsafe.As` to do the right thing, but that would require users to pull in another package to do this stuff...

I had thought about providing additional overloads (such as `UnpackLow(float, float)`), but that still results in "indeterminate" results in some scenarios (such as the two upper bits of the result).

File: tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+                    for (int i = 0; i < testsCount; i++)
+                    {
+                        Span<double> value = doubleTable.GetAssignmentData(i).Span;
+                        Vector128<double> result = Sse2.SetVector128(value[1], value[0]);
+                        doubleTable.SetOutArray(result, i);
+                    }
+
+                    if (Environment.Is64BitProcess)
+                    {
+
```
Comment (tannergooding) [inline tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs]: nit: stray newline

File: tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+                                Span<long> value = longTable.GetAssignmentData(i).Span;
+                                Vector128<long> result = Sse2.SetVector128(value[1], value[0]);
+                                longTable.SetOutArray(result, i);
+                            }
+                            testResult = Fail;
+                            Console.WriteLine($"{nameof(Sse2)}.{nameof(Sse2.SetVector128)} failed on long: expected PlatformNotSupportedException exception.");
+                        }
+                        catch (PlatformNotSupportedException)
+                        {
+
```
Comment (tannergooding) [inline tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs]: Haven't we been putting comments in the catch block for other similar tests?

File: tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+                    for (int i = 0; i < testsCount; i++)
+                    {
+                        Span<byte> value = byteTable.GetAssignmentData(i).Span;
+                        Vector128<byte> result = Sse2.SetVector128(value[15], value[14], value[13], value[12], value[11], value[10], value[9],
+                            value[8], value[7], value[6], value[5], value[4], value[3], value[2], value[1], value[0]);
+                        byteTable.SetOutArray(result, i);
+                    }
+
+                    CheckMethodSpan<double> checkDouble = (Span<double> x, Span<double> y, Span<double> z, Span<double> a) =>
```
Comment (tannergooding) [inline tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs]: What are `y` and `a` for?

File: tests/src/JIT/HardwareIntrinsics/X86/Sse2/TestTableSse2.cs [hunk 8]
```diff
@@ -224,7 +224,7 @@ public bool CheckResult(CheckMethod<T> check)
         public bool CheckResult(CheckMethodSpan<T> check)
         {
             bool result = true;
-            for (int i = 0; i < inArray1.Length; i++)
+            for (int i = 0; i < inArray1.Length; i += ElementCount)
```
Comment (tannergooding) [inline tests/src/JIT/HardwareIntrinsics/X86/Sse2/TestTableSse2.cs]: Why this change?

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs [hunk 9]
```diff
@@ -971,47 +972,191 @@ public static class Sse2
...
         /// </summary>
-        public static Vector128<double> SetVector128(double e1, double e0) => SetVector128(e1, e0);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe Vector128<double> SetVector128(double e1, double e0)
+        {
+            // TODO-CQ Optimize algorithm choice based on benchmarks
+
+            // Zero vector and load e1 et index 0
+            Vector128<double> e1Vector = LoadScalarVector128(&e1);
+            Vector128<double> e0Vector = LoadScalarVector128(&e0);
```
Comment (fiigii) [inline src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs]: I do not like these memory round-trips (arg-reg -> memory-> vector reg), `SetScalarVector128<T>` should be better.

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs (hunk 4 above)
Comment (4creators) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs]: > Maybe we could update Unsafe.As to do the right thing

Actually this what are you commenting on are some of the most serious pain points in current API surface. We need easy and low cost way to load from memory and move from register any of the supported numerals into lowest position of the vector:

```C#

int value = Foo();
.... 
Vector<int> vector = int.AsVector();
// this may allow to write
Vector<int> vector = UnapckLow(value.AsVector(), value.AsVector());
```

Comment (4creators) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs]: We could even go further and create helper methods which would expand into corresponding intrinsics like:

```C#

int value = Foo();
Vector<int> vector = value.Shuffle(0);
vector = value.UnpackLow();

// and corresponding vector methods
vector = vector.Shuffle(1);

// what would allow to write
vector = vector.ShiftLeft(4).Shuffle(3).UnpackLow();
```

File: tests/src/JIT/HardwareIntrinsics/X86/Sse2/TestTableSse2.cs (hunk 8 above)
Comment (4creators) [reply tests/src/JIT/HardwareIntrinsics/X86/Sse2/TestTableSse2.cs]: This method was not used before and was created specifically for `Span<T>` arguments. It needs to increment by element count of `Vector<T>`

File: src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs (hunk 9 above)
Comment (4creators) [reply src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs]: Yes, you are right.

File: tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs (hunk 7 above)
Comment (4creators) [reply tests/src/JIT/HardwareIntrinsics/X86/Sse2/SetVector128.cs]: I have reused some of the test API instead of creating additional methods and for this particular case reused method has too many arguments.

**Comment by 4creators** [comment]
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

**Comment by 4creators** [comment]
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

**Comment by 4creators** [comment]
test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx

**Comment by 4creators** [comment]
One unrelated failure:

```
JIT_Methodical._cctor_misc_threads2_cs_do_threads2_cs_do_._cctor_misc_threads2_cs_do_threads2_cs_do_cmd [FAIL] 
```
in  https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_jitx86hwintrinsicnoavx_prtest/198/

and timout in another job reruning tests

**Comment by 4creators** [comment]
@tannergooding It is second timeout - should I rerun?

Can we increase timeout to 5h?

```
    Finished:    GC.API.XUnitWrapper
    Finished:    Regressions.common.XUnitWrapper
    Finished:    tracing.tracevalidation.XUnitWrapper
    Finished:    JIT.superpmi.XUnitWrapper
    Finished:    Regressions.coreclr.XUnitWrapper
    Finished:    JIT.SIMD.XUnitWrapper
    Finished:    baseservices.exceptions.XUnitWrapper
    Finished:    JIT.Performance.XUnitWrapper
    Finished:    JIT.CodeGenBringUpTests.XUnitWrapper
    Finished:    JIT.opt.XUnitWrapper
    Finished:    JIT.Generics.XUnitWrapper
    Finished:    JIT.IL_Conformance.XUnitWrapper
    Finished:    JIT.HardwareIntrinsics.XUnitWrapper
    Finished:    GC.Scenarios.XUnitWrapper
    Finished:    JIT.Directed.XUnitWrapper
    Finished:    JIT.jit64.XUnitWrapper
    Finished:    baseservices.threading.XUnitWrapper
    Finished:    JIT.Regression.XUnitWrapper
Build timed out (after 240 minutes). Marking the build as aborted.
Build was aborted
``

**Comment by tannergooding** [comment]
> should I rerun

No. The HardwareIntrinsic tests completed without failure and the changes wouldn't impact the other tests.

> Can we increase timeout to 5h?

@jkotas might be able to comment on this...

**Comment by jkotas** [comment]
> Can we increase timeout to 5h?

I do not see a problem with increasing the timeout.

@jashook ?

**Comment by jashook** [comment]
Sounds good to me, opened PR #16888 to address the issue.
